Proxied token introspection implementation

jeffAlan04 opened the pull request:
Here we implement the remote token introspection.
A list of OIDC providers should be configured, as per the `oidc` [Spring profile](https://indigo-iam.github.io/v/current/docs/reference/configuration/#spring-profiles). Since the same configuration structure is applied both to remote providers allowed for token introspection and OIDC providers used for login, here we separate the two cases (backward compatible with current configuration):
- `oidc.providers[].allow-proxied-introspection = true && oidc.providers[].loginButton.visible = false`:
  allows remote proxied introspection but not login
- `oidc.providers[].allow-proxied-introspection = false && oidc.providers[].loginButton.visible = true`:
  default behavior, only allows login with external OIDC providers.

Also, the well-known endpoint response of OIDC providers is cached when doing remote proxied introspection.

The following is a brief example of one OIDC provider configured to be used to introspect tokens and not for login purposes:

```
oidc:
  providers:
  - name: iam-remote
    issuer: https://iam-remote.local.io/
    client:
      clientId: client
      clientSecret: secret
      scope: openid,profile,email
    allow-proxied-introspection: true
    loginButton:
      visible: false
```

jeffAlan04 committed:
First proxied introspection implementation

enricovianello commented:
Adding this dependency in your `iam-login-service\pom.xml` file:

```
    <dependency>
      <groupId>org.springframework.security</groupId>
      <artifactId>spring-security-oauth2-resource-server</artifactId>
    </dependency>
```
(e.g. in lines 91-94 between the other spring framework dependencies).

We can use the `OpaqueTokenIntrospector` object as it's done in Esaco [here](https://github.com/indigo-iam/esaco/blob/develop/esaco-app/src/main/java/it/infn/mw/esaco/EsacoConfiguration.java#L209-L214).

jeffAlan04 committed:
Creation of docker compose

federicaagostini committed:
Add igi_test_ca to Java trust store

federicaagostini committed:
Use iam-test-client image

federicaagostini committed:
Add README

giacomini commented on iam-login-service/src/main/java/it/infn/mw/iam/authn/oidc/DefaultRestTemplateFactory.java:
```diff
@@ -15,7 +15,15 @@
  */
 package it.infn.mw.iam.authn.oidc;
 
+import javax.net.ssl.SSLContext;
```
Perche' tutti questi import aggiuntivi se non ci sono altre modifiche al codice in questo file?

giacomini commented on iam-login-service/src/main/java/it/infn/mw/iam/config/ProxiedIntrospectionConfig.java:
```diff
@@ -0,0 +1,51 @@
+
+@Configuration
+public class ProxiedIntrospectionConfig {
+    @Bean
```
Mi pare che l'indentazione non sia uniforme con altri file

giacomini commented on iam-login-service/src/main/java/it/infn/mw/iam/core/oauth/introspection/model/DelegatingOpaqueTokenIntrospector.java:
```diff
@@ -0,0 +1,119 @@
+
+    RestTemplate restTemplate = restTemplateMapper.apply(provider.getClient());
+
+    /** 
```
Non mi pare che il commento aggiunga informazione utile, meglio eliminare

giacomini commented on iam-login-service/src/main/java/it/infn/mw/iam/core/oauth/introspection/model/DelegatingOpaqueTokenIntrospector.java:
```diff
@@ -0,0 +1,119 @@
+    String introspectionEndpoint = discovery.has(INTROSPECTION_ENDPOINT_KEY)
+        ? discovery.get(INTROSPECTION_ENDPOINT_KEY).asText()
+        : null;
+    /** 
```
Non mi pare che il commento aggiunga informazione utile, meglio eliminare

giacomini commented on iam-login-service/src/main/java/it/infn/mw/iam/core/oauth/introspection/model/DelegatingOpaqueTokenIntrospector.java:
```diff
@@ -0,0 +1,119 @@
+    * JSON
+    */
+    JsonNode discovery = discoveryService.getDiscoveryDocument(issuer, restTemplate);
+    /**
```
Non mi pare che il commento aggiunga informazione utile, meglio eliminare

giacomini commented on iam-login-service/src/main/java/it/infn/mw/iam/core/oauth/introspection/IamIntrospectionService.java:
```diff
@@ -89,15 +95,29 @@ public IntrospectionResponse introspect(Authentication auth, String tokenValue,
+            OAuth2AccessTokenEntity at = tokenService.readAccessToken(tokenValue);
+            response = introspectAccessToken(authenticatedClient, at, info);
+          } else {
+            // Calls the introspector and gets the OAuth2AuthenticatedPrincipal
```
Commento poco utile

giacomini commented on iam-login-service/src/main/java/it/infn/mw/iam/core/oauth/introspection/IamIntrospectionService.java:
```diff
@@ -89,15 +95,29 @@ public IntrospectionResponse introspect(Authentication auth, String tokenValue,
+
+            LOG.info("Introspection result for token issued by {}: active={}",
+                issuer, introspectionResult.getAttribute("active"));
+                
```
Commento poco utile

giacomini commented on iam-login-service/src/main/java/it/infn/mw/iam/core/oauth/introspection/IamIntrospectionService.java:
```diff
@@ -204,13 +229,27 @@ private boolean notYetValid(JWTClaimsSet claims) {
   }
+
+  private IntrospectionResponse convertIntrospectionResponse(OAuth2AuthenticatedPrincipal principal) {
+    // Creates a builder for costructing the IntrospectionResponse
```
Commento poco utile

giacomini commented on iam-login-service/src/main/java/it/infn/mw/iam/core/oauth/introspection/IamIntrospectionService.java:
```diff
@@ -204,13 +229,27 @@ private boolean notYetValid(JWTClaimsSet claims) {
+  private IntrospectionResponse convertIntrospectionResponse(OAuth2AuthenticatedPrincipal principal) {
+    // Creates a builder for costructing the IntrospectionResponse
+    IntrospectionResponse.Builder builder = new IntrospectionResponse.Builder(principal.getAttribute("active"));
+    /**
```
Commento poco utile

giacomini commented on iam-login-service/src/main/java/it/infn/mw/iam/core/oauth/introspection/IamIntrospectionService.java:
```diff
@@ -204,13 +229,27 @@ private boolean notYetValid(JWTClaimsSet claims) {
+     * Every claim gets copied as additional field
+     */
+    principal.getAttributes().forEach(builder::addField);
+    // Creates the final IntrospectionResponse object from the builder and returns
```
Commento poco utile

federicaagostini commented on iam-login-service/src/main/java/it/infn/mw/iam/core/oauth/discovery/DefaultOidcDiscoveryService.java:
```diff
@@ -0,0 +1,68 @@
+    for (String url : DISCOVERY_URLS) {
+      try {
+        ResponseEntity<String> resp = restTemplate.getForEntity(base + url, String.class);
+        LOG.info("Discover response: {} {}",  resp.getStatusCode(), resp.getBody());
```
I would log the response body in case the log is set to DEBUG rather than INFO. Return the status code of the introspection result with INFO is fine.

federicaagostini committed:
Address Francesco's comments

federicaagostini committed:
Remove unused files

federicaagostini committed:
Delete unused application files

federicaagostini committed:
Move oidc-dev into resources

federicaagostini committed:
Hopefully more readable code

federicaagostini committed:
Fix conflict with Test class

federicaagostini committed:
Fix issues

federicaagostini committed:
Fix tests

federicaagostini committed:
Add flag for remote introspection of OIDC providers

federicaagostini committed:
Fix test

federicaagostini committed:
Cache OIDC discovery document response

federicaagostini committed:
Add missing license

federicaagostini committed:
Fix test

federicaagostini committed:
Add OIDC discovery cache tests

federicaagostini committed:
Use property to hide OIDC providers from login

Since the same configuration structure is applied both to remote providers allowed
for token introspection and OIDC providers used for login, here we separate the two cases:
- oidc.providers[].allow-proxied-introspection = true && oidc.providers[].loginButton.visible = false:
  allows remote proxied introspection but not login
- oidc.providers[].allow-proxied-introspection = false && oidc.providers[].loginButton.visible = true:
  default behavior, only allows login with external OIDC providers.

federicaagostini committed:
Fix login button visibility in login page

enricovianello commented on iam-login-service/src/main/java/it/infn/mw/iam/config/CacheConfig.java:
```diff
@@ -40,7 +41,7 @@ CacheManager fakeCacheManager(CacheProperties props) {
   CacheManager localCacheManager(CacheProperties props) {
     return new ConcurrentMapCacheManager(IamWellKnownInfoProvider.CACHE_KEY,
-        DefaultScopeMatcherRegistry.SCOPE_CACHE_KEY);
+        DefaultScopeMatcherRegistry.SCOPE_CACHE_KEY, DefaultOidcDiscoveryService.CACHE_NAME);
```
Check if we can evict entries after a certain amount of time elapsed since the creation

enricovianello commented on iam-login-service/src/main/java/it/infn/mw/iam/config/CacheConfig.java:
```diff
@@ -50,6 +51,8 @@ RedisCacheManagerBuilderCustomizer redisCacheManagerBuilderCustomizer() {
       .withCacheConfiguration(IamWellKnownInfoProvider.CACHE_KEY,
           RedisCacheConfiguration.defaultCacheConfig())
       .withCacheConfiguration(DefaultScopeMatcherRegistry.SCOPE_CACHE_KEY,
+          RedisCacheConfiguration.defaultCacheConfig())
```
same as above

enricovianello commented on iam-login-service/src/main/java/it/infn/mw/iam/config/TaskConfig.java:
```diff
@@ -96,6 +97,13 @@ public TaskConfig(NotificationStoreService notificationStoreService,
+  
+  @Scheduled(fixedRateString = "${task.oidcDiscoveryCacheCleanupPeriodSecs:86400}",
+      timeUnit = TimeUnit.SECONDS)
+  @CacheEvict(allEntries = true, cacheNames = DefaultOidcDiscoveryService.CACHE_NAME)
```
See above Can we avoid to delete all the entries at the same time?

enricovianello commented on iam-login-service/src/test/java/it/infn/mw/iam/test/oauth/introspection/OpaqueTokenIntrospectorTests.java:
```diff
@@ -0,0 +1,240 @@
+import it.infn.mw.iam.core.oauth.introspection.model.DelegatingOpaqueTokenIntrospector;
+
+@SuppressWarnings("deprecation")
+@ActiveProfiles({"h2"})
```
I think this annotation can be removed

enricovianello commented on iam-login-service/src/test/java/it/infn/mw/iam/test/oauth/introspection/OidcDiscoveryCacheTests.java:
```diff
@@ -0,0 +1,178 @@
+
+  @Nested
+  @SpringBootTest(properties = {"cache.enabled=true", "cache.redis.enabled=false"})
+  @ActiveProfiles({"h2-test", "dev"})
```
Removing the scheduled task should allow to remove this (and similar)

enricovianello commented on iam-login-service/src/test/java/it/infn/mw/iam/test/oauth/introspection/ProxiedIntrospectionTests.java:
```diff
@@ -0,0 +1,104 @@
+
+@IamMockMvcIntegrationTest
+@SpringBootTest(classes = {IamLoginService.class}, webEnvironment = WebEnvironment.MOCK)
+@ActiveProfiles({"h2"})
```
ActiveProfiles can be removed. You could include @AutoConfigureMockMvc with no params and @Transactional from @IamMockMvcIntegrationTest and remote this annotation.

enricovianello reviewed (changes requested):
Some small changes

federicaagostini committed:
Solve some of the Enrico's comment

federicaagostini committed:
Add TTL after writing in cache

Also, switch to Caffeine cache instead of ConcurrentMapCacheManager

federicaagostini committed:
Fix Sonar issues
